### PR TITLE
[Merged by Bors] - feat(Analysis/Calculus/FDeriv/Mul): Add DifferentiableAt.finset_prod

### PR DESCRIPTION
**Pull request by CBirkbeck**
Add results saying that if a finite sequence of function is differentiable then so is their product. 

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by loefflerd** [comment]
This is about functions of a scalar argument, hence `deriv` rather than `fderiv`; so it should not go in the `fderiv` file but rather in `Analysis.Calculus.Deriv.Mul`, somewhere close to the existing lemma `HasDerivAt.finset_prod`. Here's a one-line proof of (a generalisation of) your lemma using that existing one:

```lean
import Mathlib

variable {𝕜 : Type*} [NontriviallyNormedField 𝕜] 
  {ι : Type*} [DecidableEq ι] {𝔸' : Type*} [NormedCommRing 𝔸'] [NormedAlgebra 𝕜 𝔸']
  {u : Finset ι} {f : ι → 𝕜 → 𝔸'} 

theorem DifferentiableAt.finset_prod (hd : ∀ i ∈ u, DifferentiableAt 𝕜 (f i) s) :
    DifferentiableAt 𝕜 (∏ i ∈ u, f i ·) s := 
  (HasDerivAt.finset_prod (fun i hi ↦ DifferentiableAt.hasDerivAt (hd i hi))).differentiableAt
```

(you can probably allow a slightly more general target ring as in `HasDerivAt.finset_prod`, in case anyone needs this for functions `

File: Mathlib/Analysis/Calculus/FDeriv/Mul.lean [hunk 1]
```diff
@@ -22,7 +22,7 @@ This file contains the usual formulas (and existence assertions) for the derivat
 
 
 open scoped Classical
-open Filter Asymptotics ContinuousLinearMap Set Metric Topology NNReal ENNReal
+open Filter Asymptotics ContinuousLinearMap Set Metric Topology NNReal ENNReal BigOperators
```
Comment (loefflerd) [inline Mathlib/Analysis/Calculus/FDeriv/Mul.lean:25]: git glitch. It's always a good idea to check your PR's on github to make sure you're not modifying any files you didn't mean to.

File: Mathlib/Analysis/Calculus/Deriv/Mul.lean [hunk 2]
```diff
@@ -356,6 +356,10 @@ theorem derivWithin_finset_prod (hxs : UniqueDiffWithinAt 𝕜 s x)
       ∑ i ∈ u, (∏ j ∈ u.erase i, f j x) • derivWithin (f i) s x :=
   (HasDerivWithinAt.finset_prod fun i hi ↦ (hf i hi).hasDerivWithinAt).derivWithin hxs
 
+theorem DifferentiableAt.finset_prod (t : 𝕜) (hd : ∀ i ∈ u, DifferentiableAt 𝕜 (f i) t) :
+    DifferentiableAt 𝕜 (∏ i ∈ u, f i ·) t :=
```
Comment (loefflerd) [inline Mathlib/Analysis/Calculus/Deriv/Mul.lean:360]: better to reuse existing variables for consistency with neighbourhing lemmas:
```suggestion
theorem DifferentiableAt.finset_prod (hd : ∀ i ∈ u, DifferentiableAt 𝕜 (f i) x) :
    DifferentiableAt 𝕜 (∏ i ∈ u, f i ·) x :=
```

File: Mathlib/Analysis/Calculus/Deriv/Mul.lean [hunk 3]
```diff
@@ -356,6 +356,20 @@ theorem derivWithin_finset_prod (hxs : UniqueDiffWithinAt 𝕜 s x)
...
+  (HasDerivAt.finset_prod (fun i hi ↦ DifferentiableAt.hasDerivAt (hd i hi))).differentiableAt
+
+theorem DifferentiableOn.finset_prod (hd : ∀ i ∈ u, DifferentiableOn 𝕜 (f i) s) :
+    DifferentiableOn 𝕜 (∏ i ∈ u, f i ·) s := by
+  simp_rw [DifferentiableOn] at *
+  exact fun t ht ↦ (HasDerivWithinAt.finset_prod
+  (fun i hi ↦ DifferentiableWithinAt.hasDerivWithinAt (hd i hi t ht))).differentiableWithinAt
```
Comment (loefflerd) [inline Mathlib/Analysis/Calculus/Deriv/Mul.lean:367]:
```suggestion
    DifferentiableOn 𝕜 (∏ i ∈ u, f i ·) s :=
  fun x hx ↦ (HasDerivWithinAt.finset_prod
    (fun i hi ↦ DifferentiableWithinAt.hasDerivWithinAt (hd i hi x hx))).differentiableWithinAt
```

File: Mathlib/Analysis/Calculus/Deriv/Mul.lean [hunk 4]
```diff
@@ -356,6 +356,20 @@ theorem derivWithin_finset_prod (hxs : UniqueDiffWithinAt 𝕜 s x)
...
+
+theorem Differentiable.finset_prod (hd : ∀ i ∈ u, Differentiable 𝕜 (f i)) :
+    Differentiable 𝕜 (∏ i ∈ u, f i ·):=
+  fun t => DifferentiableAt.finset_prod t (fun i hi => hd i hi t)
```
Comment (loefflerd) [inline Mathlib/Analysis/Calculus/Deriv/Mul.lean:371]:
```suggestion
theorem Differentiable.finset_prod (hd : ∀ i ∈ u, Differentiable 𝕜 (f i)) :
    Differentiable 𝕜 (∏ i ∈ u, f i ·) :=
  fun x ↦ .finset_prod x (fun i hi ↦ hd i hi x)
```

Comment (CBirkbeck) [reply Mathlib/Analysis/Calculus/Deriv/Mul.lean]: FYI I'm not intentionally starting an arrow war (=> vs ↦ ) its just that I keep automatically writing that, even though I agree that ↦  looks better!

**Comment by loefflerd** [comment]
Can you add the “DifferentiableWithinAt” version to complete the set?

**Comment by CBirkbeck** [comment]
> Can you add the “DifferentiableWithinAt” version to complete the set?

Oops yes sorry! done now :)

**Comment by loefflerd** [comment]
maintainer merge

Thanks!

**Pull request closed without merging**